components/eamxx/src/control/intensive_observation_period.cpp

```diff
@@ -237,6 +495,342 @@ read_fields_from_file_for_iop (const std::string& file_name,
   }
 }
 
+void IntensiveObservationPeriod::
+read_iop_file_data (const util::TimeStamp& current_ts)
+{
+  // If no iop file is given, return early
+  if (not m_params.isParameter("iop_file")) return;
+
+  const auto iop_file = m_params.get<std::string>("iop_file");
+  const auto iop_file_time_idx = m_time_info.get_iop_file_time_idx(current_ts);
+
+  // Sanity check
+  EKAT_REQUIRE_MSG(iop_file_time_idx >= m_time_info.time_idx_of_current_data,
+                   "Error! Attempting to read previous iop file data time index.\n");
+
+  // If we are still in the time interval as the previous read from iop file,
+  // there is no need to reload data. Return early
+  if (iop_file_time_idx == m_time_info.time_idx_of_current_data) return;
+
+  const auto file_levs = scorpio::get_dimlen(iop_file, "lev");
+  const auto iop_file_pressure = m_helper_fields["iop_file_pressure"];
+  const auto model_pressure = m_helper_fields["model_pressure"];
+  const auto surface_pressure = m_iop_fields["Ps"];
+
+  // Loop through iop fields, if rank 1 fields exist we need to
+  // gather information for vertically interpolating views
+  bool has_level_data = false;
+  for (auto& it : m_iop_fields) {
+    if (it.second.rank() == 1) {
+      has_level_data = true;
+      break;
+    }
+  }
+
+  // Compute values and indices associate with pressure for interpolating data (if necessary).
+  int adjusted_file_levs;
+  int iop_file_start;
+  int iop_file_end;
+  int model_start;
+  int model_end;
+  if (has_level_data) {
+    // Load surface pressure (Ps) from iop file
+    Real* ps_data = surface_pressure.get_view<Real, Host>().data();
+    read_variable_from_file(iop_file, "Ps", "real", {"lon","lat"}, iop_file_time_idx, ps_data);
+    surface_pressure.sync_to_dev();
+
+    // Pre-process file pressures, store number of file levels
+    // less than or equal to the surface pressure.
+    const auto iop_file_pres_v = iop_file_pressure.get_view<Real*>();
+    // Sanity check
+    EKAT_REQUIRE_MSG(file_levs+1 == iop_file_pres_v.extent_int(0),
+                    "Error! Unexpected size for helper field \"iop_file_pressure\"\n");
+    const auto& Ps = surface_pressure.get_view<const Real>();
+    Kokkos::parallel_reduce(file_levs+1, KOKKOS_LAMBDA (const int ilev, int& lmin) {
```

Comment:
What's the point of doing a parallel reduce here? It appears that only `ilev==file_levs` does anything, so why the loop?

Reply:
You're right, this computation was incorrect (but worked in this case). I've updated it.

The basic goal is to have the reduction tell us the min number of levels where the final level is equal to the surface pressure.

Reply:
Ah, I now see you also do stuff on other levels.